src/docs/development/add-to-app/android/plugin-setup.md

```diff
@@ -0,0 +1,128 @@
+
+Taking the [firebase_crashlytics plugin](https://pub.dev/packages/firebase_crashlytics)
+as an example. The plugin's integration instructions requires some manual edits
+to your Android wrapper project's `build.gradle` file.
+
+In the case of a Flutter module, which only has Dart files, perform those
+Gradle file edits on your outer, existing Android app rather than in your Flutter
+module.
+
+{{site.alert.note}}
+Astute readers may notice that the Flutter module directory also contain
+an `.android` and an `.ios` directory. Those directories are
+Flutter-tool-generated and are only meant to bootstrap Flutter into generic
+Android or iOS libraries. They should not be edited or checked-in. This
+allows Flutter to improve the integration point should there be bugs or updates
+needed with new versions of Gradle, Android, Android Gradle Plugin, etc.
+
+For advanced users, if more modularity is needed and you must not leak knowledge
+of your Flutter module's dependencies into your outer host app, you can rewrap
+and repackage your Flutter module's Gradle library inside another native
+Android Gradle library that depends on the Flutter module's Gradle library.
+You can make your Android specific changes such as editing the
+AndroidManifest.xml, Gradle files or adding more Java files in that wrapper
+library.
+{{site.alert.end}}
+
+## C. Merging libraries
+
+The scenario which requires slightly more attention is if your existing Android
+application already depends on the same Android library your Flutter module
+does (transitively via a plugin).
+
+For instance, your existing app's Gradle may already have:
+
+<!--code-excerpt "<existing app>/app/build.gradle" title-->
+```gradle
+…
+dependencies {
+  …
+  implementation 'com.crashlytics.sdk.android:crashlytics:2.10.1'
+  …
+}
+…
+```
+
+And your Flutter module also depends on [firebase_crashlytics](https://pub.dev/packages/firebase_crashlytics)
+via pubspec.yaml:
+
+<!--code-excerpt "<Flutter module>/pubspec.yaml" title-->
+```yaml
+…
+dependencies:
+  …
+  firebase_crashlytics: ^0.1.3
+  …
+…
+```
+
+This plugin usage will transitively add a Gradle dependency again via
+firebase_crashlytics v0.1.3's own [Gradle file](https://github.com/FirebaseExtended/flutterfire/blob/bdb95fcacf7cf077d162d2f267eee54a8b0be3bc/packages/firebase_crashlytics/android/build.gradle#L40):
+
+<!--code-excerpt "<firebase_crashlytics via pub>/android/build.gradle" title-->
+```gradle
+…
+dependencies {
+  …
+  implementation 'com.crashlytics.sdk.android:crashlytics:2.9.9'
+  …
+}
+…
+```
+
+The two `com.crashlytics.sdk.android:crashlytics` dependencies may not be the same
+version. In this example, the host app requested v2.10.1 and the Flutter
+module plugin requested v2.9.9.
+
+By default, Gradle v5 will [resolve dependency version conflicts](https://docs.gradle.org/current/userguide/dependency_resolution.html#sub:resolution-strategy)
+by using the newest version of the library.
```

Comment:
> by using the newest version of the library

It picks the first dependency that shows up in Graph while traversing. :/

Reply:
Even they admit that this is bad. "With Maven, this could have unwanted impact on resolved versions.".

Reply:
lol is that right? 

https://docs.gradle.org/current/userguide/dependency_resolution.html says "Gradle will consider all requested versions, wherever they appear in the dependency graph. Out of these versions, it will select the highest one."

But it's so random I wouldn't be surprised if the implementation did that too.

Reply:
I guess even the docs is confusing. It also says:

> Apache Maven uses a nearest first strategy.
>
> Maven will take the shortest path to a dependency and use that version. In case there are multiple paths of the same length, the first one wins.